src/scalecube-transport/provider/RSocketProvider.js

```diff
@@ -0,0 +1,98 @@
+import RSocketWebSocketClient from 'rsocket-websocket-client';
+import WebSocket from 'isomorphic-ws';
+import { JsonSerializers, RSocketClient } from 'rsocket-core';
+import { Observable } from 'rxjs';
+import { validateRequest } from '../utils';
+
+export class RSocketProvider {
+
+  constructor() {
+    this._client = null;
+    this._socket = null;
+  }
+
+  build({ url, keepAlive = 60000, lifetime = 180000, wsCreator = url => new WebSocket(url) }) {
+    return new Promise((resolve, reject) => {
+      this._client = new RSocketClient({
+        serializers: JsonSerializers,
+        setup: {
+          keepAlive,
+          lifetime,
+          dataMimeType: 'application/json',
+          metadataMimeType: 'application/json',
+        },
+        transport: new RSocketWebSocketClient({ url, wsCreator }),
+      });
+      this._connect()
+        .then(resolve)
+        .catch(reject)
+    });
+  }
+
+  _connect() {
+    return new Promise((resolve, reject) => {
+      this._client.connect().subscribe({
+        onComplete: (socket) => {
+          this._socket = socket;
+          resolve();
+        },
+        onError: reject
+      });
+    });
+  }
+
+  _disconnect() {
+    return new Promise((resolve, reject) => {
+      if (!this._socket) {
+        return reject('The connection is not opened');
+      }
+      this._socket = null;
+      this._client.close();
+      resolve();
+    });
+  }
+
+  request(requestData) {
+    const { headers: { type, responsesLimit }, data, entrypoint } = requestData;
+    const isSingle = type === 'requestResponse';
+    const isStream = type === 'requestStream' || type === 'requestChannel';
+    const initialRespondsAmount = responsesLimit || 1;
+
+    return Observable.create((subscriber) => {
+      const validationError = validateRequest(requestData);
+      if (validationError) {
+        return subscriber.error(new Error(validationError));
+      }
+
+      let unsubscribe;
+      let socketSubscriber;
+      this._socket[type]({ data, metadata: { q: entrypoint }})
+        .subscribe({
+          onNext: (response) => {
+            subscriber.next(response.data);
+            !responsesLimit && socketSubscriber && socketSubscriber.request(1);
```

Comment:
I read little bit about it after taking with Ronen
.request(n), is for you to control what coming to avoid overflow
From what I am understands, it just how many you item you want to take for processing right now
You will still get item from server but won't emit them to the subscribe processing the stream
 
https://github.com/rsocket/rsocket-js/blob/230f3d01a08317ea949ce42eaf2012c6e1e25a5a/packages/rsocket-types/src/ReactiveStreamTypes.js
https://github.com/rsocket/rsocket-js/blob/dfc40d6e58dbb0f2cc7e21b80e91837a20a2cfa0/docs/03-flowable-api.md

Reply:
I don't understand your point here.
I run it in the browser with request(3) and received only 3 emitions in network and no indicator that the stream is completed.
I've reproduced the same behavior in the test **_Use requestStream type with "many" action with responsesLimit = 4 - receive 4 responses and the stream is not completed_**

Reply:
@dmitriyslaym please discuss this with @artem-v he will better explain it on site